### ISSUE TITLE
ReadH5AD and Jerry-Rigging the Seurat Object to run FindAllMarkers

### ISSUE THREAD

**Issue by MediciPrime** [satijalab/seurat#1199]
I have recently used a machine learning [tool](https://eleozzr.github.io/desc/) to cluster my cells however that tool doesn't have an equivalent FindAllMarkers function from Seurat.  I went through that tool's [tutorial](https://eleozzr.github.io/desc/tutorial.html) and then converted the AnnData object to a Seurat object using ReadH5AD.  The DESC pipeline isn't a one-to-one representation of the Seurat pipeline so ReadH5AD justifiably complained about some missing information.  

Missing Information:  
`No feature loadings found`
`Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to EMBEDED_`
`Warning: No columnames present in cell embeddings, setting to 'EMBEDED_1:16'`
`Warning: No columnames present in cell embeddings, setting to 'tSNE_1:2'`
`Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to TSNE1_`
`Warning: No columnames present in cell embeddings, setting to 'TSNE1_1:2'`
`Warning: No columnames present in cell embeddings, setting to 'UMAP_1:2'`
`Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to UMAP1_`
`Warning: No columnames present in cell embeddings, setting to 'UMAP1_1:2'`
`Assembling Seurat object`
`No nearest-neighbor graph`

The newly created Seurat object does have a complete meta.data table including cell 'barcode' information and the clusters belonging to the cells.  I just want to identify the marker genes that define each of the clusters but I can't use FindAllMarkers because my Seurat object is missing some information.  

For example, one of the problems I can identify is that the ident or 'active.ident' table wasn't properly created since only one factor 'ctrl' is present for all the cell barcodes.  I was thinking about manually fixing the 'active.ident' table in order to get FindAllMarkers to work but I have a feeling that I will likely run into additional errors.  Mainly because when I currently run 'FindAllMarkers' the following errors are displayed:  
`Warning: No DE genes identified`
`Warning: The following tests were not performed:`
`Cell group 2 is empty - no cells with identity class`
`data frame with 0 columns and 0 rows`

I understand that this isn't a typical problem but is there any way I can manually fix the Seurat object in order to get FindAllMarkers to work?

**Comment by andrewwbutler** [comment]
What are the identities currently set to? You can check this with `Idents(object)`. If this doesn't correspond to the clusters you identified, you can set this with `Idents(object) <- "metadata_columnname"` and then rerun the differential expression analysis.

**Comment by eleozzr** [comment]
When you want to convert an AnnData after running you can do by,
``` R
# assume adata is the object after running DESC clustering procedure
suppressMessages(library(Seurat))
suppressMessages(library(reticulate))
#  Sys.setenv(RETICULATE_PYTHON="/usr/bin/python3") # set the path of your python tool
ad=import("anndata",convert = FALSE)# Note: convert=FALSE is necessay.
adata=ad$read("adata_desc.h5ad") 
adata=change_obs(adata,exclude = c("n_gene","n_counts","tsne1","tsne2")) # change obs change non-numeric column to string, exclue parameter specify the column(s) that are not converted, please fin
obj=Convert_from_anndata_to_seurat(adata,raw.X.slot = "count.data")
obj=NormalizeData(obj,display.progress = F)
#then you can set ident to find marker genes
obj=SetAllIdent(obj,id="desc_1.0")# Also, you can set other meta.data variable
markers=FindAllMarkers(obj)
```
The following two functions are necessary. 

#Sys.setenv(RETICULATE_PYTHON="/usr/bin/python3") # set the path of your python tool
Convert_from_anndata_to_seurat=function(from=adata,X.slot="scale.data",raw.X.slot="logcount.data"){
	if(!py_module_available("anndata")) {
	        stop("Please install the anndata python module")
	 }
	stopifnot(X.slot%in%c("scale.data","normlizecount.data"))
	stopifnot(raw.X.slot%in%c("count.data","logcount.data"))
	data.matrix=tryCatch(
                expr=t(py_to_r(from$X)),
               	        error=function(e){
                       	        stop("No adata.X in provided adata. If Both adata.X and adata.raw.X are None")
                       	}
       )
	rownames(data.matrix)<-rownames(py_to_r(from$var))
	colnames(data.matrix)<-rownames(py_to_r(from$obs))	
	if (X.slot=="normalizecount.data") X.slot="data"

	raw.data.matrix <- tryCatch(
        	expr = t(py_to_r(from$raw$X)),
	        error = function(e) {
        		stop("No adata.raw.X in provided adata. please make sure adata have adata.raw.X when you tyr to turn it to `seurat object`")
	        }
      	)
	if(raw.X.slot=="logcount.data"){
		raw.data.matrix=expm1(raw.data.matrix)
	}
	
	rownames(x = raw.data.matrix) <- rownames(x = py_to_r(from$raw$var))
	colnames(x = raw.data.matrix) <- rownames(x =py_to_r(from$obs))
	#get meta.data
	meta.data=py_to_r(from$obs)
	if ("nUMI" %in% colnames(x = meta.data)) {
		colnames(x = meta.data) <- gsub(
		pattern = "nUMI",
		replacement = "nUMI_ori",
		x = colnames(x = meta.data)
		)
	}
	if ("nGene" %in% colnames(x = meta.data)) {
		colnames(x = meta.data) <- gsub(
		pattern = "nGene",
		replacement = "nGene_ori",
		x = colnames(x = meta.data)
		)
	}
	seurat.object <- CreateSeuratObject(raw.data = raw.data.matrix,meta.data = meta.data)
	seurat.object <- SetAssayData(
        	object = seurat.object,
	        assay.type = "RNA",
	        slot = X.slot,
	        new.data = data.matrix
        )
	#deal with obsm fields that are not dimensional reductions, or have different name structures
	drs <- unlist(x = py_to_r(from$obsm$keys()))
	for (dr in drs) {
        	dr.embed <- py_to_r(from$obsm[[eval(dr)]])
	        

change_obs=function(adata,exclude=c("n_genes","n_counts")){
  n_col=as.numeric(as.character(adata$obs$columns$shape[0]))
  tmp=adata$obs$columns
  for (i in seq_len(n_col)){
    cur_colname=as.character(tmp[i-1])
    if (!cur_colname %in% exclude){
      adata$obs[[cur_colname]]=adata$obs[[cur_colname]]$astype("str")
    }
  }
  return(adata)
}

**Comment by MediciPrime** [comment]
Thank you @andrewwbutler and @eleozzr for your help!  
@andrewwbutler I was able to manually fix the Seurat object by specifying the correct clusters for the cells identified by DESC using:

`Idents(object) <- "desc.0.8"`

This allowed me to use `FindAllMarkers` flawlessly, thank you!

@eleozzr I tried out your method and it also worked.  Pretty awesome that the creator of DESC was also able to help me out with my problem, thank you!
